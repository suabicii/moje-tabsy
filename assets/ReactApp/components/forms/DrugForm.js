import React, {useState} from "react";
import DosingMomentsInputs from "./DosingMomentsInputs";

function DrugForm(props) {
    const dosingMomentInputDefaultAmount = props.drug ? Object.keys(props.drug.dosingMoments).length : 1;
    const dosingMoments = props.drug ? Object.entries(props.drug.dosingMoments) : null;
    const [dosingMomentInputAmount, setDosingMomentInputAmount] = useState(dosingMomentInputDefaultAmount);
    const handleChangeDosingInput = value => {
        const dosingMomentInputs = [...document.querySelectorAll('.dosing-moment-input')];
        if (value > dosingMomentInputs.length) {
            setDosingMomentInputAmount(dosingMomentInputAmount + 1);
        } else {
            setDosingMomentInputAmount(dosingMomentInputAmount - 1);
        }
    };

    return (
        <>
            <div className="card-header text-center">
                <button className="btn btn-close float-start" onClick={() => {
                    props.setIsFormVisible(false);
                }}></button>
                {props.drug ? 'Edytuj' : 'Dodaj'} lek/suplement
            </div>
            <div className="card-body">
                <form role="form" name="drug_form" onSubmit={e => {
                    e.preventDefault();
                }}>
                    <div className="row">
                        <div className="col-md-6">
                            <div className="form-floating mt-3">
                                <input type="text" className="form-control" id="name" name="name"
                                       defaultValue={props.drug ? props.drug.name : ""} placeholder="Nazwa" required
                                />
                                <label htmlFor="name">Nazwa</label>
                            </div>
                            <div className="form-floating mt-3">
                                <input type="text" className="form-control" id="unit" name="unit"
                                       defaultValue={props.drug ? props.drug.unit : ""} placeholder="Jednostka" required
                                />
                                <label htmlFor="unit">Jednostka (szt., ml. itp.)</label>
                            </div>
                            <div className="form-floating mt-3">
                                <input type="number" step="0.01" name="dosing" id="dosing" className="form-control"
                                       min="0.01"
                                       aria-valuemin="0.01"
                                       defaultValue={props.drug ? props.drug.dosing : ""}
                                       placeholder="Dawkowanie (ile na raz?)" required
                                />
                                <label htmlFor="dosing">Dawkowanie (ile na raz?)</label>
                            </div>
                        </div>
                        <div className="col-md-6">
                            <div className="form-floating mt-3">
                                <input type="number" step="1" className="form-control" id="quantity" name="quantity"
                                       defaultValue={props.drug ? props.drug.quantity : ""} placeholder="Bieżąca ilość"
                                       min="1" aria-valuemin="1" required
                                />
                                <label htmlFor="quantity">Bieżąca ilość</label>
                            </div>
                            <div className="form-floating mt-3">
                                <input type="number" step="1" className="form-control" id="quantityMax"
                                       defaultValue={props.drug ? props.drug.quantityMax : ""} name="quantityMax"
                                       placeholder="Ilość całkowita" min="1" aria-valuemin="1" required
                                />
                                <label htmlFor="quantityMax">Ilość całkowita</label>
                            </div>
                            <div className="form-floating mt-3">
                                <input type="number" step="1" className="form-control" id="dailyDosing"
                                       name="daily_dosing"
                                       placeholder="Ile razy dziennie?"
                                       defaultValue={props.drug ? Object.keys(props.drug.dosingMoments).length : "1"}
                                       aria-valuemin="1" min="1"
                                       required data-testid="dailyDosing"
                                       onChange={e => {
                                           handleChangeDosingInput(e.target.value);
                                       }}
                                />
                                <label htmlFor="dailyDosing">Ile razy dziennie?</label>
                            </div>
                        </div>
                    </div>
                    <div className="dosing-hours mt-3">
                        <h4 className="text-center">Godziny przyjęcia dawki</h4>
                        <DosingMomentsInputs dosingMoments={dosingMoments} inputAmount={dosingMomentInputAmount}/>
                    </div>
                    <div className="d-flex justify-content-between mt-3">
                        <button type="submit" className="btn btn-info px-5">Zapisz</button>
                        <button type="button" className="btn btn-outline-dark px-5" onClick={() => {
                            props.setIsFormVisible(false);
                        }}>
                            Anuluj
                        </button>
                    </div>
                </form>
            </div>
        </>
    );
}

export default DrugForm;