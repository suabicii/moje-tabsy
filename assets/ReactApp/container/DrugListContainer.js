import React, {useState} from "react";
import {createContainer} from "unstated-next";

const useDrugListing = (initialState = []) => {
    const [drugList, setDrugList] = useState(initialState);
    const addDrug = ({id, name, quantity, quantityMax, unit, dosing, dosingMoments}) => setDrugList([
        ...drugList,
        {
            id,
            name,
            quantity,
            quantityMax,
            unit,
            dosing,
            dosingMoments
        }
    ]);
    const removeDrug = id => setDrugList(drugList.filter(drug => drug.id !== id));
    const editDrug = (id, updates = {}) => setDrugList(drugList.map(drug => drug.id === id ? {...drug, ...updates} : drug));
    return {drugList, addDrug, removeDrug, editDrug};
};

export const DrugListContainer = createContainer(useDrugListing);