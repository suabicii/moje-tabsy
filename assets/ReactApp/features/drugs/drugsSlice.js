import {createSelector, createSlice} from "@reduxjs/toolkit";
import {fetchData} from "../../utils/fetchData";

const drugsSlice = createSlice({
    name: 'drugs',
    initialState: [],
    reducers: {
        addDrug: (state, action) => {
            const {id, name, quantity, quantityMax, unit, dosing, dosingMoments} = action.payload;
            state.push({id, name, quantity, quantityMax, unit, dosing, dosingMoments});
        },
        removeDrug: (state, action) => {
            const id = action.payload;
            return state.filter(drug => drug.id !== id);
        },
        editDrug: (state, action) => {
            const {id, ...updates} = action.payload;
            const drugIndex = state.findIndex(drug => drug.id === id);
            if (drugIndex !== -1) {
                state[drugIndex] = {...state[drugIndex], ...updates};
            }
        },
        setDrugs: (state, action) => action.payload
    },
    sort: true
});

const {reducer, actions} = drugsSlice;
export const {addDrug, removeDrug, editDrug, setDrugs} = actions;

const defaultDrugsSelector = state => state.drugs;

export const sortedDrugsSelector = createSelector(
    defaultDrugsSelector,
    drugs => [...drugs].sort((a, b) => a.name.localeCompare(b.name))
);

export const fetchDrugs = () => async dispatch => {
    await fetchData('/api/drug-list').then(data => {
        dispatch(setDrugs(data));
    }).catch(err => {
        console.log(err);
    });
};

export default reducer;