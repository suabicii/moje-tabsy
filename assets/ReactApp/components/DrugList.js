import React, {useEffect, useState} from "react";
import {mainRoute} from "../routers/AppRouter";
import {useNavigate} from "react-router-dom";
import {DrugListContainer} from "../container/DrugListContainer";
import DrugForm from "./forms/DrugForm";

function DrugList(props) {
    const navigate = useNavigate();
    const {drugList, removeDrug} = DrugListContainer.useContainer();
    const [isEditFormVisible, setIsEditFormVisible] = useState(false);
    const [isAddFormVisible, setIsAddFormVisible] = useState(false);
    const [drugDataForEdit, setDrugDataForEdit] = useState(undefined);

    useEffect(() => {
        localStorage.setItem('drugs', JSON.stringify(drugList));
    });

    return (
        <>
            <div className="card text-center mt-3">
                <div className="card-header">
                    Leki/suplementy <i className="fa-solid fa-pills"></i>
                </div>
                <div className="card-body">
                    <h5 className="card-title">Lista leków i suplementów:</h5>
                    <ul className="card-text list-group">
                        {drugList.map(drug => <li key={drug.id} className="list-group-item" data-testid="drug">
                                <strong>{drug.name}</strong>: {drug.quantity}/{drug.quantityMax} {`${drug.unit}, `}
                                <strong>Dzienna
                                    dawka: </strong> {drug.dosing} {drug.unit} {`${Object.keys(drug.dosingMoments).length} raz(-y) dziennie `}
                                {
                                    // Delete button
                                    props.isEditMode &&
                                    <button className="btn btn-danger rounded-circle float-md-end"
                                            data-testid={`remove-drug-${drug.id}`} onClick={() => {
                                        removeDrug(drug.id);
                                    }}>
                                        <i className="fa-solid fa-trash-can"></i>
                                    </button>
                                }
                                {
                                    // Edit button
                                    props.isEditMode &&
                                    <button className="btn btn-info rounded-circle float-md-end"
                                            data-testid={`edit-drug-${drug.id}`} onClick={() => {
                                        setIsEditFormVisible(true);
                                        setDrugDataForEdit(drug);
                                    }}>
                                        <i className="fa-solid fa-pencil"></i>
                                    </button>
                                }
                            </li>
                        )}
                    </ul>
                    <div className="d-grid">
                        <button className="btn btn-primary mt-2" data-testid="add-drug" onClick={() => {
                            if (props.isEditMode) {
                                setIsAddFormVisible(true);
                            } else {
                                navigate(`${mainRoute}/drug-list`);
                            }
                        }}>
                            {props.isEditMode ? 'Dodaj ' : 'Edytuj '}
                            {props.isEditMode ? <i className="fa-solid fa-plus"></i> :
                                <i className="fa-solid fa-pencil"></i>}
                        </button>
                    </div>
                </div>
            </div>
            {
                props.isEditMode &&
                <>
                    {
                        isAddFormVisible &&
                        <div className="card mt-3">
                            <DrugForm setIsFormVisible={setIsAddFormVisible}/>
                        </div>
                    }
                    {
                        isEditFormVisible &&
                        <div className="card mt-3">
                            <DrugForm drug={drugDataForEdit} setIsFormVisible={setIsEditFormVisible}/>
                        </div>
                    }
                </>
            }
        </>
    );
}

export default DrugList;