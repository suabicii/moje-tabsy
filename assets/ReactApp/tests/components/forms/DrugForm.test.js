/**
 * @jest-environment jsdom
 * */
import React from "react";
import ReactShallowRenderer from "react-test-renderer/shallow";
import DrugForm from "../../../components/forms/DrugForm";
import {fireEvent, render, screen} from "@testing-library/react";
import drugs from "../fixtures/drugs";
import {DrugListContainer} from "../../../container/DrugListContainer";
import {BrowserRouter} from "react-router-dom";
import DrugList from "../../../components/DrugList";

const renderForm = (drug = null) => {
    const setIsFormVisible = jest.fn();
    if (drug) {
        render(
            <DrugListContainer.Provider initialState={drugs}>
                <DrugForm drug={drug} setIsFormVisible={setIsFormVisible}/>
            </DrugListContainer.Provider>
        );
    } else {
        render(
            <DrugListContainer.Provider initialState={drugs}>
                <DrugForm setIsFormVisible={setIsFormVisible}/>
            </DrugListContainer.Provider>
        );
    }
};

it('should correctly render DrugForm', () => {
    const renderer = new ReactShallowRenderer();
    const setIsFormVisible = jest.fn();
    renderer.render(
        <DrugListContainer.Provider>
            <DrugForm setIsFormVisible={setIsFormVisible}/>
        </DrugListContainer.Provider>
    );
    expect(renderer.getRenderOutput).toMatchSnapshot();
});

it('should render drug edit form with drug data', () => {
    renderForm(drugs[0]);

    expect(screen.getByRole('form')).toHaveFormValues({
        name: drugs[0].name,
        unit: drugs[0].unit,
        quantity: drugs[0].quantity,
        quantityMax: drugs[0].quantityMax,
        dosing: drugs[0].dosing,
        daily_dosing: Object.keys(drugs[0].dosingMoments).length,
        hour1: drugs[0].dosingMoments["hour1"],
        hour2: drugs[0].dosingMoments["hour2"]
    });
});

it('should render modal with empty input fields in most and one default value after clicking add button', () => {
    renderForm();
    expect(screen.getByRole('form')).toHaveFormValues({daily_dosing: 1});
});

it('should add dosing moment input after increasing daily dosing input value in add drug form', () => {
    renderForm();

    fireEvent.change(screen.getByTestId('dailyDosing'), {target: {value: 2}});

    expect(screen.getAllByRole('timer').length).toBeGreaterThan(1);
});

it('should remove dosing moment input added before after decreasing daily dosing input value in add drug form', () => {
    renderForm();

    fireEvent.change(screen.getByTestId('dailyDosing'), {target: {value: 2}});
    fireEvent.change(screen.getByTestId('dailyDosing'), {target: {value: 1}});

    expect(screen.getAllByRole('timer').length).toBe(1);
});

it('should add dosing moment input after increasing more than once daily dosing input value in edit drug form', () => {
    renderForm(drugs[0]);

    fireEvent.change(screen.getByTestId('dailyDosing'), {target: {value: 2}});
    fireEvent.change(screen.getByTestId('dailyDosing'), {target: {value: 3}});

    expect(screen.getAllByRole('timer').length).toBeGreaterThan(2);
});

it('should remove dosing moment input after decreasing daily dosing input value in edit drug form', () => {
    renderForm(drugs[0]);

    fireEvent.change(screen.getByTestId('dailyDosing'), {target: {value: 3}});
    fireEvent.change(screen.getByTestId('dailyDosing'), {target: {value: 2}});

    expect(screen.getAllByRole('timer').length).toBe(2);
});

it('should change dosing moment input amount to correct value after manually writing any daily dosing input value any times in edit drug form', () => {
    renderForm(drugs[0]);

    fireEvent.change(screen.getByTestId('dailyDosing'), {target: {value: 8}});
    fireEvent.change(screen.getByTestId('dailyDosing'), {target: {value: 4}});

    expect(screen.getAllByRole('timer').length).toBe(4);
});

it('should remove dosing moment inputs with values from database after decreasing daily dosing input value in edit drug form', () => {
    renderForm(drugs[0]);

    fireEvent.change(screen.getByTestId('dailyDosing'), {target: {value: 1}}); // decreased value from value equaled 2

    expect(screen.getAllByRole('timer').length).toBe(1);
});

it('should add drug to list', () => {
    const root = document.createElement('div');
    document.body.appendChild(root);
    render(
        <DrugListContainer.Provider initialState={drugs}>
            <BrowserRouter>
                <DrugList isEditMode={true} customRoot={root}/>
            </BrowserRouter>
        </DrugListContainer.Provider>
    );
    const drugListLengthBeforeUpdate = screen.getAllByTestId('drug').length;

    fireEvent.click(screen.getByTestId('add-drug')); // activate add drug form

    fireEvent.change(screen.getByTestId('drugName'), {target: {value: 'Prozac'}});
    fireEvent.change(screen.getByTestId('unit'), {target: {value: 'pcs'}});
    fireEvent.change(screen.getByTestId('dosing'), {target: {value: '1'}});
    fireEvent.change(screen.getByTestId('quantity'), {target: {value: '120'}});
    fireEvent.change(screen.getByTestId('quantityMax'), {target: {value: '120'}});
    fireEvent.change(screen.getByTestId('dailyDosing'), {target: {value: '1'}});
    fireEvent.change(screen.getByTestId('hour1-test'), {target: {value: '07:00'}});
    fireEvent.submit(screen.getByRole('form'));

    const drugListLengthAfterUpdate = screen.getAllByTestId('drug').length;

    expect(drugListLengthAfterUpdate).toBeGreaterThan(drugListLengthBeforeUpdate);
});