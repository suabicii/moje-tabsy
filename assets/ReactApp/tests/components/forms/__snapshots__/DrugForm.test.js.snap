// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should correctly render DrugForm 1`] = `[Function]`;
